src/VideoReceiver/GStreamer.cc

```diff
@@ -108,17 +108,37 @@ static void qgcputenv(const QString& key, const QString& root, const QString& pa
 static void
 blacklist()
 {
-    GstRegistry* reg;
+    GstRegistry* registry = gst_registry_get();
 
-    if ((reg = gst_registry_get()) == nullptr) {
+    if (registry == nullptr) {
+        qCCritical(GStreamerLog) << "Failed to get gstreamer registry.";
         return;
     }
 
-    GstPluginFeature* plugin;
+    auto changeRank = [registry](const char* featureName, uint16_t rank) {
+        GstPluginFeature* feature = gst_registry_lookup_feature(registry, featureName);
+        if (feature == nullptr) {
+            qCWarning(GStreamerLog) << "Failed to change ranking of feature:" << featureName;
+            return;
+        }
+
+        qCInfo(GStreamerLog) << "Changing feature (" << featureName << ") to use rank:" << rank;
+        gst_plugin_feature_set_rank(feature, rank);
+        gst_registry_add_feature(registry, feature);
+        gst_object_unref(feature);
+    };
+
+    // Set rank for specific features
+    changeRank("bcmdec", GST_RANK_NONE);
+
+    // Enable VAAPI drivers
+    for(auto name : {"vaapimpeg2dec", "vaapimpeg4dec", "vaapih263dec", "vaapih264dec", "vaapivc1dec", "vaapidecodebin"}) {
+        changeRank(name, GST_RANK_PRIMARY);
+    }
 
-    if ((plugin = gst_registry_lookup_feature(reg, "bcmdec")) != nullptr) {
-        qCCritical(GStreamerLog) << "Disable bcmdec";
-        gst_plugin_feature_set_rank(plugin, GST_RANK_NONE);
+    // Enable NVIDIA's proprietary APIs for hardware video acceleration
+    for(auto name : {"nvh265dec", "nvh265sldec", "nvh264dec", "nvh264sldec"}) {
```

Comment:
Just came over https://gstreamer.freedesktop.org/documentation/nvcodec/nvh264dec.html?gi-language=c and found its rank is 'primary'. So why we need to assign GST_RANK_PRIMARY to nvh264dec again? Aren't other elements also 'primary' already? Is not, then, is not this extra option in the QGC settings useless?

Reply:
@patrickelectric looked on rank of a few HW decoders and seems that a lot of them use GST_RANK_PRIMARY in contrast others which are GST_RANK_PRIMARY + 1. The best solution would be to iterate over hw decoders and give them all rank GST_RANK_PRIMARY + 1 or work in reverse way - call this option 'lower software decoding priority' and set avdec_h264 rank to GST_RANK_PRIMARY - 1 or similar. But only if this options is enabled.

Reply:
@andrewvoznytsa I'm doing some tests based on this idea, but also with a second approach where I could filter all hardware decodes and their rank and let the user choose how to configure it.

We got a weird problem where gstreamer was trying to use vaapi, that mostly has higher ranks compared to nvdpau, with the proprietary nvidia driver, resulting in a non working pipeline.

I'm going to do more tests and share here.

Reply:
@patrickelectric perhaps it would be even better to let advanced users to choose which decider to blacklist and which to make higher priority - just to list and one 'advanced' option. And leave default gstreamer ranks for most of users. It is not trivial task to find config which works good for everyone.

Reply:
@andrewvoznytsa that's exactly what I'm planning to do.

Reply:
I think this has been addressed by forcing the decoder to the user selection, without changing the default behavior.